[MSE] Remove SourceBufferParser::Segment.

jyavenard opened the pull request:
#### b98e5e1c6baa53edf0e7437baedab9f06e4dbd58
<pre>
[MSE] Remove SourceBufferParser::Segment.
<a href="https://bugs.webkit.org/show_bug.cgi?id=243370">https://bugs.webkit.org/show_bug.cgi?id=243370</a>
<a href="https://rdar.apple.com/97831219">rdar://97831219</a>

Reviewed by Youenn Fablet.

We remove the SourceBufferParser::Segment class and directly deal with
SharedBuffer.
To better control lifetime of the segment, we immediately move in SourceBufferParser::appendData
the rvalue Ref&lt;const SharedBuffer&gt;

Follow-up on 288710@main which fixed long standing issue in
WebMParser::SegmentReader::Read where if the data parsed was across two
segments, we could have returned corrupted data (this was a theoretical issue that couldn&apos;t be reproduced from JS).

No functional changes, covered by existing tests.

* Source/WebCore/platform/audio/cocoa/AudioFileReaderCocoa.cpp:
(WebCore::AudioFileReader::demuxWebMData const):
* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferParserAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferParserAVFObjC.mm:
(WebCore::SourceBufferParserAVFObjC::appendData):
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
(WebCore::MediaPlayerPrivateWebM::dataReceived):
* Source/WebCore/platform/graphics/cocoa/SourceBufferParser.cpp:
(WebCore::SourceBufferParser::Segment::Segment): Deleted.
(WebCore::SourceBufferParser::Segment::size const): Deleted.
(WebCore::SourceBufferParser::Segment::read const): Deleted.
(WebCore::SourceBufferParser::Segment::takeSharedBuffer): Deleted.
(WebCore::SourceBufferParser::Segment::getData const): Deleted.
* Source/WebCore/platform/graphics/cocoa/SourceBufferParser.h:
* Source/WebCore/platform/graphics/cocoa/SourceBufferParserWebM.cpp:
(WebCore::WebMParser::parse):
(WebCore::SourceBufferParserWebM::appendData):
(WebCore::segmentReadErrorToWebmStatus): Deleted.
* Source/WebCore/platform/graphics/cocoa/SourceBufferParserWebM.h:

Canonical link: <a href="https://commits.webkit.org/296589@main">https://commits.webkit.org/296589@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8465c382e1bd412cbf22782bd747e01ee8631af5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/108971 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/28631 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/19056 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/114180 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/59306 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/110934 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/29314 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/37198 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/82808 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/111919 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/23309 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/98154 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/63247 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/22728 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/16289 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/58877 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/92684 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/16337 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/117299 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/36020 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/26625 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/91817 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/36392 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/94416 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/91625 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23325 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/36537 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/14293 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/31885 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/35918 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/41441 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/35619 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/38959 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/37303 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferParserAVFObjC.mm:
```diff
@@ -241,11 +241,11 @@ String extractCodecName(AVAssetTrack* track)
     INFO_LOG_IF_POSSIBLE(LOGIDENTIFIER);
-    auto sharedBuffer = segment.takeSharedBuffer();
-    auto nsData = sharedBuffer->makeContiguous()->createNSData();
+    Ref sharedBuffer = WTFMove(segment);
```
Why do we need to move segment?
Can we do:
`RetainPtr nsData = segment->createNSData()`?
Maybe we could pass a `const&` depending on the other implementations

jyavenard replied:
To control the lifecycle and ensuring the segment is deleted on the work queue.

youennf commented on Source/WebCore/platform/graphics/cocoa/SourceBufferParserWebM.cpp:
```diff
@@ -533,12 +522,13 @@ void WebMParser::reset()
-ExceptionOr<int> WebMParser::parse(SourceBufferParser::Segment&& segment)
+ExceptionOr<int> WebMParser::parse(Ref<const SharedBuffer>&& segment)
 {
+    Ref sharedBuffer = WTFMove(segment);
```
Why is it needed?

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/49c3727f938f68a6ca922405b15eb2352d0b327e. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46954

jyavenard committed:
[MSE] Remove SourceBufferParser::Segment.
https://bugs.webkit.org/show_bug.cgi?id=243370
rdar://97831219

Reviewed by Youenn Fablet.

We remove the SourceBufferParser::Segment class and directly deal with
SharedBuffer.
To better control lifetime of the segment, we immediately move in SourceBufferParser::appendData
the rvalue Ref<const SharedBuffer>

Follow-up on 288710@main which fixed long standing issue in
WebMParser::SegmentReader::Read where if the data parsed was across two
segments, we could have returned corrupted data (this was a theoretical issue that couldn't be reproduced from JS).

No functional changes, covered by existing tests.

* Source/WebCore/platform/audio/cocoa/AudioFileReaderCocoa.cpp:
(WebCore::AudioFileReader::demuxWebMData const):
* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferParserAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferParserAVFObjC.mm:
(WebCore::SourceBufferParserAVFObjC::appendData):
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
(WebCore::MediaPlayerPrivateWebM::dataReceived):
* Source/WebCore/platform/graphics/cocoa/SourceBufferParser.cpp:
(WebCore::SourceBufferParser::Segment::Segment): Deleted.
(WebCore::SourceBufferParser::Segment::size const): Deleted.
(WebCore::SourceBufferParser::Segment::read const): Deleted.
(WebCore::SourceBufferParser::Segment::takeSharedBuffer): Deleted.
(WebCore::SourceBufferParser::Segment::getData const): Deleted.
* Source/WebCore/platform/graphics/cocoa/SourceBufferParser.h:
* Source/WebCore/platform/graphics/cocoa/SourceBufferParserWebM.cpp:
(WebCore::WebMParser::parse):
(WebCore::SourceBufferParserWebM::appendData):
(WebCore::segmentReadErrorToWebmStatus): Deleted.
* Source/WebCore/platform/graphics/cocoa/SourceBufferParserWebM.h:

Canonical link: https://commits.webkit.org/296589@main

webkit-commit-queue commented:
Committed 296589@main (b98e5e1c6baa): <https://commits.webkit.org/296589@main>

Reviewed commits have been landed. Closing PR #46954 and removing active labels.
